vscode-lean4/src/infoview.ts

```diff
@@ -314,30 +318,52 @@ export class InfoProvider implements Disposable {
                 // infoview updates properly.
                 await this.onClientRestarted(client);
             }),
+            client.restartedWorker(async () => {
+                await this.onWorkerRestarted(client);
+            }),
             client.didSetLanguage(() => this.onLanguageChanged()),
         );
 
         // Note that when new client is first created it still fires client.restarted
         // event, so all onClientRestarted can happen there so we don't do it twice.
     }
 
+    async onWorkerRestarted(client: LeanClient) : Promise<void> {
+        await this.webviewPanel?.api.serverStopped(undefined); // clear any server stopped state
+        const folder = client.getWorkspaceFolder()
+        if (this.clientsFailed.has(folder)) {
+            this.clientsFailed.delete(folder) // delete from failed clients
+            console.log('Restarting server for workspace: ' + folder)
+        }
+        await this.sendPosition();
+    }
+
     onClientRemoved(client: LeanClient) {
         // todo: remove subscriptions for this client...
     }
 
-    async onActiveClientStopped(client: LeanClient, activeClient: boolean, msg: string) {
+    async onActiveClientStopped(client: LeanClient, activeClient: boolean, msg: string, reason: string, restartFile: boolean = false) {
         // Will show a message in case the active client stops
         // add failed client into a list (will be removed in case the client is restarted)
         if (activeClient)
         {
             // means that client and active client are the same and just show the error message
-            await this.webviewPanel?.api.serverStopped(msg);
+            await this.webviewPanel?.api.serverStopped({message:msg, reason});
         }
 
-        console.log(`client stopped: ${client.getWorkspaceFolder()}`)
-
         // remember this client is in a stopped state
-        this.clientsFailed.set(client.getWorkspaceFolder(), msg)
+        const key = client.getWorkspaceFolder();
```

Comment:
This will still show an error message for *all* files in the same directory, right?  We only want to show an error message for the file that crashed.  (Maybe I'm wrong, didn't try it out.)

Reply:
Good catch...

Reply:
yeah you're right, fixing it